packages/askar/src/storage/__tests__/AskarStorageService.test.ts

```diff
@@ -0,0 +1,311 @@
+      const expectedRecord = await insertRecord({ tags: { myTag: 'foo', anotherTag: 'bar' } })
+      await insertRecord({ tags: { myTag: 'notfoobar' } })
+
+      const records = await storageService.findByQuery(agentContext, TestRecord, {
+        $and: [{ myTag: 'foo' }, { anotherTag: 'bar' }],
+      })
+
+      expect(records.length).toBe(1)
+      expect(records[0]).toEqual(expectedRecord)
+    })
+
+    it('finds records using $or statements', async () => {
+      const expectedRecord = await insertRecord({ tags: { myTag: 'foo' } })
+      const expectedRecord2 = await insertRecord({ tags: { anotherTag: 'bar' } })
+      await insertRecord({ tags: { myTag: 'notfoobar' } })
+
+      const records = await storageService.findByQuery(agentContext, TestRecord, {
+        $or: [{ myTag: 'foo' }, { anotherTag: 'bar' }],
+      })
+
+      expect(records.length).toBe(2)
+      expect(records).toEqual(expect.arrayContaining([expectedRecord, expectedRecord2]))
+    })
+
+    it('finds records using $not statements', async () => {
+      const expectedRecord = await insertRecord({ tags: { myTag: 'foo' } })
+      // FIXME: Seems to be an issue with Askar WQL implementation:
+      // it only takes into account records containing a myTag and returns only those different from the $not rule
+      // (i.e. expectedRecord2 will not be taken into account because it does not have any myTag value)
+      const expectedRecord2 = await insertRecord({ tags: { anotherTag: 'bar' } })
+      await insertRecord({ tags: { myTag: 'notfoobar' } })
+
+      const records = await storageService.findByQuery(agentContext, TestRecord, {
+        $not: { myTag: 'notfoobar' },
+      })
+
+      expect(records.length).toBe(2)
+      expect(records).toEqual(expect.arrayContaining([expectedRecord, expectedRecord2]))
+    })
+
+    it('correctly transforms an advanced query into a valid WQL query', async () => {
+      const storageService = new AskarStorageService<TestRecord>()
+
+      const expectedQuery = {
+        $and: [
+          {
+            $and: undefined,
+            $not: undefined,
+            $or: [
+              { myTag: '1', $and: undefined, $or: undefined, $not: undefined },
+              { myTag: '0', $and: undefined, $or: undefined, $not: undefined },
+            ],
+          },
+          {
+            $or: undefined,
+            $not: undefined,
+            $and: [
+              { theNumber: 'n__0', $and: undefined, $or: undefined, $not: undefined },
+              { theNumber: 'n__1', $and: undefined, $or: undefined, $not: undefined },
+            ],
+          },
+        ],
+        $or: [
+          {
+            'aValue:foo': '1',
+            'aValue:bar': '1',
+            $and: undefined,
+            $or: undefined,
+            $not: undefined,
+          },
+        ],
+        $not: { myTag: 'notfoobar', $and: undefined, $or: undefined, $not: undefined },
+      }
+
+      expect(
+        //@ts-ignore
```

Comment:
Why the ts-ignore? Should we add an eslint disable here?

Reply:
This is because I'm calling a private method. Is there an elegant way to call a private method from jest? I know the correct way of doing things is to test it indirectly but in this case just wanted to check the internals.

Reply:
Yeah sometimes that's fine. It sometimes add typing with a comment:

```
// we want to test the private method
type A = Original & { privateMethod: something }
const a = b as unknown as A
```

An alternative is to extract the logic into an util and testing the util.

Your approach is fine as well, but a comment would be good!

Reply:
Thanks for the info! I moved them to an util to make it cleaner.